js/zb.js

```diff
@@ -1185,29 +1261,58 @@ module.exports = class zb extends Exchange {
     }
 
     sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let url = this.urls['api'][api];
-        if (api === 'public') {
-            url += '/' + this.version + '/' + path;
-            if (Object.keys (params).length) {
-                url += '?' + this.urlencode (params);
-            }
-        } else if (api === 'trade') {
+        const [ section, version, access ] = api;
+        let url = this.urls['api'][section][version][access];
+        if (path === 'getFeeInfo') {
             url += '/' + path;
             if (Object.keys (params).length) {
                 url += '?' + this.urlencode (params);
             }
-        } else {
-            let query = this.keysort (this.extend ({
-                'method': path,
-                'accesskey': this.apiKey,
-            }, params));
-            const nonce = this.nonce ();
-            query = this.keysort (query);
-            const auth = this.rawencode (query);
-            const secret = this.hash (this.encode (this.secret), 'sha1');
-            const signature = this.hmac (this.encode (auth), this.encode (secret), 'md5');
-            const suffix = 'sign=' + signature + '&reqTime=' + nonce.toString ();
-            url += '/' + path + '?' + auth + '&' + suffix;
+        } else if (section === 'spot') {
+            if (access === 'public') {
+                url += '/' + version + '/' + path;
+                if (Object.keys (params).length) {
+                    url += '?' + this.urlencode (params);
+                }
+            } else {
+                let query = this.keysort (this.extend ({
+                    'method': path,
+                    'accesskey': this.apiKey,
+                }, params));
+                const nonce = this.nonce ();
+                query = this.keysort (query);
+                const auth = this.rawencode (query);
+                const secret = this.hash (this.encode (this.secret), 'sha1');
+                const signature = this.hmac (this.encode (auth), this.encode (secret), 'md5');
+                const suffix = 'sign=' + signature + '&reqTime=' + nonce.toString ();
+                url += '/' + path + '?' + auth + '&' + suffix;
+            }
+        } else if (section === 'contract') {
+            if (version === 'v1') {
+                url += '/' + version + '/' + path;
+                if (Object.keys (params).length) {
+                    url += '?' + this.urlencode (params);
+                }
+            } else {
+                if (access === 'public') {
+                    url += '/' + version + '/' + path;
+                    if (Object.keys (params).length) {
+                        url += '?' + this.urlencode (params);
+                    }
```

Comment:
Ok, now, if you look into all those `'public'` sections – they're all the same, so we can reduce that down even further:

```JavaScript
if (path === 'getFeeInfo') {
    url += '/' + path;
    if (Object.keys (params).length) {
        url += '?' + this.urlencode (params);
    }
} else if (access === 'public') {
    url += '/' + version + '/' + path;
    if (Object.keys (params).length) {
        url += '?' + this.urlencode (params);
    }
} else if (section === 'spot') {
    let query = this.keysort (this.extend ({
        'method': path,
        'accesskey': this.apiKey,
    }, params));
    const nonce = this.nonce ();
    query = this.keysort (query);
    const auth = this.rawencode (query);
    const secret = this.hash (this.encode (this.secret), 'sha1');
    const signature = this.hmac (this.encode (auth), this.encode (secret), 'md5');
    const suffix = 'sign=' + signature + '&reqTime=' + nonce.toString ();
    url += '/' + path + '?' + auth + '&' + suffix;
} else if (section === 'contract') {
    let query = this.keysort (this.extend ({
        'method': path,
        'accesskey': this.apiKey,
    }, params));
    const nonce = this.nonce ();
    query = this.keysort (query);
    const auth = this.rawencode (query);
    const secret = this.hash (this.encode (this.secret), 'sha1');
    const signature = this.hmac (this.encode (auth), this.encode (secret), 'md5');
    const suffix = 'sign=' + signature + '&reqTime=' + nonce.toString ();
    url += '/' + version + '/' + path + '?' + auth + '&' + suffix;
}
```

And then, the `spot`-private and the `contract`-private sections are also very similar, so we can reduce that down to this:

```JavaScript
if (path === 'getFeeInfo') {
    url += '/' + path;
    if (Object.keys (params).length) {
        url += '?' + this.urlencode (params);
    }
} else if (access === 'public') {
    url += '/' + version + '/' + path;
    if (Object.keys (params).length) {
        url += '?' + this.urlencode (params);
    }
} else {
    let query = this.keysort (this.extend ({
        'method': path,
        'accesskey': this.apiKey,
    }, params));
    const nonce = this.nonce ();
    query = this.keysort (query);
    const auth = this.rawencode (query);
    const secret = this.hash (this.encode (this.secret), 'sha1');
    const signature = this.hmac (this.encode (auth), this.encode (secret), 'md5');
    const suffix = 'sign=' + signature + '&reqTime=' + nonce.toString ();
    if (section === 'contract') {
        url += '/' + version;
    }
    url += '/' + path + '?' + auth + '&' + suffix;
}
```